### DISCUSSION TITLE
tar_read() with capabilities like `format = "file"`?

### DISCUSSION THREAD

**Discussion by joelnitta** [ropensci/targets#785] [Help] [answered]
I have a project where I am loading objects (built targets) from another target project with `tar_read`. For example,

```
plastome_tree = tar_read(
    plastome_tree,
    store = ftol_cache
  )
```
where `ftol_cache` is the path to the `_targets` store of the other project (here called "ftol").

The problem is, if the object in the "ftol" `_targets` store  changes, the current project doesn't know about that and doesn't invalidate. It would be nice if there were an archetype like `tar_file()` that detected this and re-loads it as necessary. Does that exist?

**Comment by joelnitta** [comment, marked as answer]
10 minutes later I realized a work-around: just track the "ftol" `_targets` store with `tar_file` and use that as a dynamic target:

```
tar_file(
  ftol_cache,
  "path/to/other/store/_targets"
)
```

Perhaps not ideal though as **any** change to that store will cause it to invalidate, not just for the targets I'm interested in.

**Reply by wlandau** [reply]
You could extend the path to `"_targets/objects/plastome_tree"` and it will only track that one file.

**Reply by joelnitta** [reply]
Yes, that would work for the example I gave... but another problem is that I have several of these to load, which I now realize wasn't clear from my example.

I think this could be a good situation for a target factory (to avoid two targets per object loaded with `tar_read()`). I may try to come up with one myself for my own project, but it could possibly be useful in tarchetypes if this is a common enough use-case (no idea about that!)

**Reply by wlandau** [reply]
You could use something like  `tarchetypes::tar_files_input(..., list.files("_targets/objects", pattern = "plastome_tree_targets_"))`

**Reply by joelnitta** [reply]
Sorry, I realized I am missing something here.

The below approach works because `ftol_cache` is a target **and** is a path to a data store:

```
tar_file(
  ftol_cache,
  "path/to/other/store/_targets"
),

plastome_tree = tar_read(
    plastome_tree,
    store = ftol_cache
  )
```

If I were to change `tar_file()` to only track one file within the cache, I couldn't use that target as the argument to `store` of `tar_read()`.

I suppose one work-around would be to make a wrapper for `tar_read()` (call it `tar_read_extra()`) that accepts a dummy variable with `...` and use that to enforce a dependency on a single file. Is that what you had in mind?

```
tar_file(
  plastome_tree_file,
  "path/to/other/store/_targets/objects/plastome_tree"
),

plastome_tree = tar_read_extra(
    plastome_tree,
    store = "path/to/other/store/_targets",
    ... = plastome_tree_file
  )
```

**Reply by wlandau** [reply]
I'm not sure I follow, but you could do something like this:

```r
plastome_tree = tar_read(
  plastome_tree,
  store = tar_read(ftol_cache)
)
```

> I suppose one work-around would be to make a wrapper for tar_read() (call it tar_read_extra()) that accepts a dummy variable with ... and use that to enforce a dependency on a single file.

Dependency relationships come into play for running or visualizing the whole pipeline, but `tar_read()` works on a target-by-target basis and does not need to be aware of upstream dependencies.

**Reply by joelnitta** [reply]
I think we are talking past each other a bit, but I found [a better example in your manual](https://books.ropensci.org/targets/projects.html#sharing-targets) anyways.

In order to do what I want, I would just modify this code

```
library(targets)
source("R/functions_data.R")
source("R/functions_analysis.R")
source("R/functions_visualization.R")
tar_option_set(packages = "tidyverse")
list(
  tar_target(file_from_project_a, "store_a/objects/target_name", format = "file"),
  tar_target(data_from_project_a, readRDS(file_from_project_a)), # Assumes format = "rds" in project A.
  tar_target(new_target, analyze_data(data_from_project_a)),
  ...
)
```

to this:

```
library(targets)
library(tarchetypes)
source("R/functions_data.R")
source("R/functions_analysis.R")
source("R/functions_visualization.R")
tar_option_set(packages = "tidyverse")
list(
  tar_file_read(
     data_from_project_a,
    "store_a/objects/target_name",
    readRDS(file = !!.x) # Assumes format = "rds" in project A.
  ),
  tar_target(new_target, analyze_data(data_from_project_a)),
  ...
)
```

This way, the two targets per object from the other pipeline are now reduced to one each. So the answer is: a new target factory is not needed. We can use `tar_file_read()` as it is.
